File: java-sdk/sdk/src/main/java/com/cisco/tiedie/clients/DataReceiverClient.java
// Copyright (c) 2023, Cisco Systems, Inc. and/or its affiliates.
// All rights reserved.
// See LICENSE file in this distribution.
// SPDX-License-Identifier: Apache-2.0

package com.cisco.tiedie.clients;

import com.cisco.tiedie.auth.Authenticator;
import com.cisco.tiedie.dto.telemetry.DataSubscription;
import com.cisco.tiedie.utils.ObjectMapperSingleton;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;

import lombok.NonNull;

import org.eclipse.paho.client.mqttv3.IMqttMessageListener;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;

import java.util.List;
import java.util.Locale;
import java.util.ArrayList;
import java.util.Collections;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * This class is used to receive telemetry data from the Tiedie controller.
 * 
 * <pre>
 * {@code
 * InputStream caInputStream = new FileInputStream("ca.pem");
 * Authenticator authenticator = ApiKeyAuthenticator.create(caInputStream, "app_id", "api_key");
 *
 * DataReceiverClient dataReceiverClient = new DataReceiverClient("ssl://<host>:8883", authenticator);
 * }
 * </pre>
 * <p>
 * The above example uses an {@link com.cisco.tiedie.auth.ApiKeyAuthenticator}.
 *
 * @see Authenticator
 */
public class DataReceiverClient {
    private final MqttClient mqttClient;
    private final MqttConnectOptions mqttConnectOptions;

    private final ObjectMapper mapper = ObjectMapperSingleton.getCborInstance();

    /**
     * Create a {@link DataReceiverClient} object.
     *
     * @param baseUrl       The TieDie MQTT broker URL.
     * @param authenticator {@link Authenticator} object that describes the
     *                      authentication mechanism used.
     * @throws Exception if any MQTT problem was encountered
     */
    public DataReceiverClient(String baseUrl, Authenticator authenticator) throws Exception {
        MemoryPersistence persistence = new MemoryPersistence();

        this.mqttClient = new MqttClient(baseUrl, authenticator.getClientID(), persistence);

        this.mqttConnectOptions = new MqttConnectOptions();
        mqttConnectOptions.setAutomaticReconnect(true);
        mqttConnectOptions.setCleanSession(true);

        authenticator.setAuthOptions(mqttConnectOptions);

        // For non-TLS broker URIs (e.g. tcp://), Paho rejects SSL socket factories.
        if (!isTlsBrokerUrl(baseUrl)) {
            mqttConnectOptions.setSocketFactory(null);
            mqttConnectOptions.setHttpsHostnameVerificationEnabled(false);
        }
    }

    /**
     * Connect to the MQTT broker
     *
     * @throws Exception if any MQTT problem was encountered
     */
    public void connect() throws Exception {
        if (mqttClient.isConnected()) {
            return;
        }
        mqttClient.connect(mqttConnectOptions);
    }

    /**
     * Disconnect from the MQTT broker.
     *
     * @throws Exception if any MQTT problem was encountered
     */
    public void disconnect() throws Exception {
        if (!mqttClient.isConnected()) {
            return;
        }
        mqttClient.disconnect();
    }

    /**
     * @return true if the MQTT client is connected.
     */
    public boolean isConnected() {
        return mqttClient.isConnected();
    }

    /**
     * Subscribe to a topic. The topic must be registered by the control app
     * and this data app has to be registered on that topic for the subscription
     * to be authorized.
     *
     * @param topic    The topic to subscribe on
     * @param callback A callback function that produces a {@link DataSubscription}
     *                 object
     *                 that has the subscription data
     * @throws Exception if any MQTT problem was encountered
     */
    public void subscribe(@NonNull String topic, Consumer<DataSubscription> callback) throws Exception {
        mqttClient.subscribe(topic, (receivedTopic, message) -> {
            byte[] payload = message.getPayload();
            List<DataSubscription> dataSubscriptions = decodeDataSubscriptions(payload);
            for (DataSubscription dataSubscription : dataSubscriptions) {
                callback.accept(dataSubscription);
            }
        });
    }

    /**
     * Subscribe to a topic. The topic must be registered by the control app
     * and this data app has to be registered on that topic for the subscription
     * to be authorized.
     *
     * @param topics   The topics to subscribe on
     * @param callback A callback function that produces a {@link DataSubscription}
     *                 object
     *                 that has the subscription data
     * @throws Exception if any MQTT problem was encountered
     */
    public void subscribe(@NonNull List<String> topics, BiConsumer<DataSubscription, String> callback)
            throws Exception {
        var topicsArray = topics.toArray(new String[0]);

        // create mqtt listener array for each topic that call the callback function
        var mqttListeners = new IMqttMessageListener[topicsArray.length];
        for (int i = 0; i < topicsArray.length; i++) {
            mqttListeners[i] = (receivedTopic, message) -> {
                byte[] payload = message.getPayload();
                List<DataSubscription> dataSubscriptions = decodeDataSubscriptions(payload);
                for (DataSubscription dataSubscription : dataSubscriptions) {
                    callback.accept(dataSubscription, receivedTopic);
                }
            };
        }

        mqttClient.subscribe(topicsArray, mqttListeners);
    }

    /**
     * Unsubscribe from a topic.
     *
     * @param topic topic to unsubscribe from
     * @throws Exception if any MQTT problem was encountered
     */
    public void unsubscribe(@NonNull String topic) throws Exception {
        if (!mqttClient.isConnected()) {
            return;
        }
        mqttClient.unsubscribe(topic);
    }

    /**
     * Unsubscribe from multiple topics.
     *
     * @param topics topics to unsubscribe from
     * @throws Exception if any MQTT problem was encountered
     */
    public void unsubscribe(@NonNull List<String> topics) throws Exception {
        if (!mqttClient.isConnected() || topics.isEmpty()) {
            return;
        }
        mqttClient.unsubscribe(topics.toArray(new String[0]));
    }

    private static boolean isTlsBrokerUrl(String baseUrl) {
        if (baseUrl == null) {
            return false;
        }

        String normalized = baseUrl.trim().toLowerCase(Locale.ROOT);
        return normalized.startsWith("ssl://")
                || normalized.startsWith("tls://")
                || normalized.startsWith("wss://")
                || normalized.startsWith("mqtts://");
    }

    private List<DataSubscription> decodeDataSubscriptions(byte[] payload) throws Exception {
        if (payload == null || payload.length == 0) {
            return Collections.emptyList();
        }

        JsonNode root = mapper.readTree(payload);
        if (root == null || root.isNull()) {
            return Collections.emptyList();
        }

        if (root.isArray()) {
            return mapper.convertValue(root, new TypeReference<List<DataSubscription>>() {});
        }

        List<DataSubscription> subscriptions = new ArrayList<>(1);
        subscriptions.add(mapper.treeToValue(root, DataSubscription.class));
        return subscriptions;
    }
}

File: python-sdk/tiedie/api/data_receiver_client.py
#!python
# Copyright (c) 2023, Cisco Systems, Inc. and/or its affiliates.
# All rights reserved.
# See LICENSE file in this distribution.
# SPDX-License-Identifier: Apache-2.0

"""

The mmodule defines a Python client using Paho MQTT for receiving and 
handling data from an MQTT broker, particularly for IoT applications.

"""

from typing import Any, Callable
import paho.mqtt.client as mqtt
import cbor2
from .auth import Authenticator


class DataReceiverClient:
    """ class DataReceiverClient """

    def __init__(self,
                 host: str,
                 authenticator: Authenticator,
                 port: int = 8883,
                 disable_tls: bool = False,
                 insecure_tls: bool = False):
        self.host = host
        self.port = port
        self.authenticator = authenticator
        self.mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2,
            authenticator.get_client_id(), True)
        self.authenticator.set_auth_options_mqtt(self.mqtt_client, disable_tls, insecure_tls)
        self.mqtt_client.on_connect = self.__on_connect
        self.mqtt_client.on_disconnect = self.__on_disconnect
        self.mqtt_client.on_message = self.__on_message
        self.connected = False

    def connect(self):
        """ Connect to the MQTT broker """
        self.mqtt_client.connect(self.host, self.port, 60)
        self.mqtt_client.loop_start()

    def disconnect(self):
        """ Disconnect from the MQTT broker """
        self.mqtt_client.disconnect()
        self.mqtt_client.loop_stop()

    def subscribe(self, topic: str,
                  callback: Callable[[Any], None]):
        """ Subscribe to a topic and register a callback function.

        Args:
            topic (str): The topic to subscribe to.
            callback (Callable[[Any], None]): 
                A callback function to be called when a message is received.
        """
        def on_message(_client, _userdata, msg):
            payload = msg.payload
            # data_subscription = data_app_pb2.DataSubscription()
            # data_subscription.ParseFromString(payload)
            print("Received message: ", len(payload), payload.hex())
            data = cbor2.loads(payload)
            callback(data)

        self.mqtt_client.subscribe(topic, qos=0)
        self.mqtt_client.message_callback_add(topic, on_message)

    def unsubscribe(self, topic: str):
        """ Unsubscribe from a topic.

        Args:
            topic (str): The topic to unsubscribe from.
        """
        self.mqtt_client.unsubscribe(topic)

    def __on_message(self, _userdata, message):
        print("Received message: " + str(message.payload))

    def __on_connect(self, _client, _userdata, _connect_flags, reason_code, _properties):
        if reason_code == 0:
            print("Connected to broker")
            self.connected = True
        else:
            print("Connection failed with error code " + str(reason_code))

    def __on_disconnect(self, _client, _userdata, _flags, _reason_code, _properties):
        print("Disconnected from broker")
        self.connected = False
